j5/components/gpio_pin.py

```diff
@@ -152,36 +152,38 @@ def mode(self, pin_mode: PinMode) -> None:
         if isinstance(pin_mode, GPIOPinMode):
             self._backend.set_gpio_pin_mode(self._identifier, pin_mode)
 
+    def digital_write(self, state: bool) -> None:
+        """Set the digital state of the pin."""
+        self._require_pin_modes({GPIOPinMode.DIGITAL_OUTPUT})
+        self._backend.write_gpio_pin_digital_state(self._identifier, state)
+
     @property
-    def digital_state(self) -> bool:
+    def last_digital_write(self) -> bool:
+        """
+        Get the last set digital state of the pin.
+
+        This does not perform a read operation, it only gets the last set
+        value, which is usually cached in memory.
+        """
+        self._require_pin_modes({GPIOPinMode.DIGITAL_OUTPUT})
+        return self._backend.get_gpio_pin_digital_state(self._identifier)
+
+    def digital_read(self) -> bool:
         """Get the digital state of the pin."""
         self._require_pin_modes({
-            GPIOPinMode.DIGITAL_OUTPUT,
             GPIOPinMode.DIGITAL_INPUT,
             GPIOPinMode.DIGITAL_INPUT_PULLUP,
             GPIOPinMode.DIGITAL_INPUT_PULLDOWN},
         )
 
-        # Behave differently depending on the hardware mode.
-        if self.mode is GPIOPinMode.DIGITAL_OUTPUT:
-            return self._backend.get_gpio_pin_digital_state(self._identifier)
-
         return self._backend.read_gpio_pin_digital_state(self._identifier)
 
-    @digital_state.setter
-    def digital_state(self, state: bool) -> None:
-        """Set the digital state of the pin."""
-        self._require_pin_modes({GPIOPinMode.DIGITAL_OUTPUT})
-        self._backend.write_gpio_pin_digital_state(self._identifier, state)
-
-    @property
-    def analogue_value(self) -> float:
+    def analogue_read(self) -> float:
         """Get the scaled analogue reading of the pin."""
         self._require_pin_modes({GPIOPinMode.ANALOGUE_INPUT})
         return self._backend.read_gpio_pin_analogue_value(self._identifier)
 
-    @analogue_value.setter
-    def analogue_value(self, new_value: float) -> None:
+    def analogue_write(self, new_value: float) -> None:
```

Comment:
I feel like it would be in the spirit of this PR to move the pwm functionality here into a separate `pwm_write` method - what do you think?

Reply:
It could also be useful to have `last_analogue_write` and `last_pwm_write` methods to go with `last_digital_write`, though these would need modifications to the backends and might be a bit too out of scope.

Reply:
b6b5cf6

Reply:
> It could also be useful to have `last_analogue_write` and `last_pwm_write` methods to go with `last_digital_write`, though these would need modifications to the backends and might be a bit too out of scope.

I agree with this idea, it does require changes to the backend though, so will open a separate issue.